### ISSUE TITLE
Constraints for each waypoints

### ISSUE THREAD

**Issue by arntanguy** [humanoid-path-planner/hpp-spline#6]
Hi,

I am looking into using this library to control the swing foot trajectory of a humanoid robot walk. Our MPC is continuously computing new desired footstep position as the robot is stepping, and thus the trajectory needs to be updated continuously by taking into account:

1. The starting pose (position/orientation) of the swing foot trajectory: fixed with zero initial velocity
2. The ending pose: always changing, desired to be reached with zero velocity
3. A mid-point between the starting position and end position at a given height: also changing since the end point is changing

From looking at the API, it seems that it is only possible to enforce constraints on the starting and end-point of a trajectory, but not on its intermediary waypoints. In my case, I would need to recompute the whole trajectory as the robot is stepping by taking into account the new current foot position, and expected desired footstep (end-point for the trajectory).

Is it possible to achieve this behaviour?

**Comment by stonneau** [comment]
Hi, you can use exact cubic splines to go through the waypoints that you want.
However, this might be a bad idea, as you won't be able to control the acceleration and velocity of the curve. If the problem is too constrained, you will end up with solutions that require unachievable velocities.

If you want to make sure you satisfy your constraints while reaching your waypoints, one way is to use Linear or Quadratic Programming with Bezier curves.

Using the De Casteljau algorithm you could constrain your curve to reach specific points while guaranteeing continuity. Then you can set up a number of free control points that you could optimize in your problem formulation.

**Comment by arntanguy** [comment]
Hi,
Thanks for the information, I'll try to find the time to look at it and integrate it for our ICRA submission.
When we last met, you mentioned that you had an example of a similar curve-generation problem? If it's relevant, it'd be helpful to see a similar concrete example.

Best regards,
Arnaud

**Comment by stonneau** [comment]
take a look at 
https://gitlab.com/stonneau/bezier_COM_traj

In particular https://gitlab.com/stonneau/bezier_COM_traj/blob/master/include/bezier-com-traj/solve_end_effector.hh 
However, the code for generating effector trajectories is not refactored yet, so it might be a bit hard to follow. The overall idea is to solve a QP problem to place one or more control points on the curve.

Do not hesitate to ask Pierre about this
